Mathlib/GameTheory/Auction/Secondprice.lean

```diff
@@ -0,0 +1,290 @@
+  let ⟨a, b, h⟩ := exists_pair_ne α
+  exact ⟨a, Finset.mem_univ _, b, Finset.mem_univ _, h⟩
+
+/-- `B i` is the maximal bid of all participants but `i`. -/
+noncomputable def B (i : a.I) : ℝ := Finset.sup' (Finset.erase Finset.univ i)
+(Finset.Nontrivial.erase_nonempty (Finset.univ_nontrivial)) b
+
+/--The second highest bid: the highest bid excluding the winner’s bid.-/
+noncomputable def secondPrice : ℝ := B b (winner b)
+
+/-- The bid of the winner is always greater than or equal to the bids of all other participants.-/
+lemma b_winner_max (j : a.I) : b j ≤ b (winner b) := by
+  rw [winner_take_max b]
+  exact Finset.le_sup' b (Finset.mem_univ j)
+
+/-- If `i`'s bid is higher than all other bids, then `i` is the winner. -/
+lemma gt_wins (i : a.I) (H : ∀ j , i ≠ j → b j < b i) : i = winner b := by
+  contrapose! H
+  exact ⟨winner b, H, b_winner_max b i⟩
+
+/-- The bid of the winner is always greater than or equal to the second highest bid. -/
+lemma b_winner : secondPrice b ≤ b (winner b) := by
+  rw [winner_take_max b]
+  exact Finset.sup'_mono b (Finset.subset_univ _)
+    (Finset.Nontrivial.erase_nonempty (Finset.univ_nontrivial))
+
+/-- If `i` is not the winner, then the highest bid excluding `i` is equal to the highest bid. -/
+lemma b_loser_max {i : a.I} (H : i ≠ winner b) : B b i = maxb b := by
+  have H1: B b i ≤ maxb b := by
+    apply Finset.sup'_le
+    intro i _
+    apply Finset.le_sup'
+    simp only [Finset.mem_univ]
+  have H2: maxb b ≤ B b i := by
+    rw [← winner_take_max b]
+    apply Finset.le_sup'
+    simp only [Finset.mem_univ, Finset.mem_erase, and_true]
+    exact (H.symm)
+  exact Real.partialOrder.proof_4 (B b i) (Finset.sup' Finset.univ maxb.proof_1 b) H1 H2
+
+namespace secondPrice
+
+/-- Defines the utility of participant `i`,
+which is their valuation minus the second highest bid if `i` is the winner, otherwise, it's 0. -/
+noncomputable def utility (i : a.I) : ℝ := if i = winner b then a.v i - secondPrice b else 0
+
+/-! ### Proofs and Lemmas -/
+variable {i : a.I}
+/-- If `i` is the winner, then their utility is their valuation minus the second highest bid. -/
+lemma utility_winner (H: i = winner b) : utility b i = a.v i - secondPrice b:= by
+  rw [utility]; simp only [ite_true, H]
+
+/-- If `i` is not the winner, then their utility is 0. -/
+lemma utility_loser (i : a.I) (H : i≠ winner b) : utility b i = 0 := by
+  rw [utility]; simp only [ite_false, H]
+
+/-- utility is non-negative if the bid equals the valuation. -/
+lemma utility_nneg (i : a.I) (H : b i = a.v i) : 0 ≤ utility b i := by
+  by_cases H2: i = winner b
+  · rw [utility]
+    simp only [H2]
+    rw [← H2, ← H, H2]
+    rw [winner_take_max b]
+    apply sub_nonneg.mpr
+    rw [secondPrice]
+    apply Finset.sup'_le
+    simp only [Finset.mem_univ, Finset.mem_erase, and_true]
+    intro j _
+    rw [maxb]
+    simp only [Finset.le_sup'_iff, Finset.mem_univ, true_and]
+    use j
+  · rw[utility, if_neg H2]
```

Comment:
Here's a golf:
```suggestion
  rcases eq_or_ne i (winner b) with rfl | H2
  · rw [utility, if_pos rfl, ← H, winner_take_max b, sub_nonneg, secondPrice]
    apply Finset.sup'_mono
    exact Finset.erase_subset (winner b) Finset.univ
  · rw [utility, if_neg H2]
```

Reply:
The fact `B b (winner b) ≤ maxb b` gets used in both `b_loser_max` and `utility_nneg`. Should it be extracted to it's own lemma?

Reply:
Nice suggestion!